### ISSUE TITLE
Tasks

### ISSUE THREAD

**Issue by lyolikaa** [aws/amazon-ecs-agent#2050]

**Comment by lyolikaa** [comment]
Hi there!
We have a Fargate cluster with some (1 or more) tasks on it, and deploying Docker image
After that we expect updated environment will run in Running tasks
I haven't found any setting to restart them automatically or on a condition
Sure , we can stop task (it will recreated), but it is risky to lose some Running data

Thanks in advance
Oleksii

**Comment by adnxn** [comment]
@lyolikaa have you looked at running [services](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs_services.html)? it's an abstraction to help manage task lifecycles on ecs and fargate.

**Comment by AndreyAttr** [comment]
@adnxn, 
Try to add some details. 
**Preconditions**:
- we uploaded new docker image to ECR,
- we have service in our cluster and saying 2 tasks running into our service. Each task is set onto ECR **latest** image and code of our image is running infinitely(loop). I mean we start each task once and forever. We need it to monitor our SQS queue every 30 seconds(or even faster in nearest future).
**Change**: Now we need to do changes in our code and we upload those changes with new image to ECR.
**Questions**:
1) Is there a safe way that our already existing & running tasks finished they work correctly and then re-started with changes from new image? I don't know how it was, but I see this at least once: I uploaded new image(just simple changes in logging) and checked in CloudWatch logs the tasks started to log with those changes. If that works this way all the time, it fully solve our problem. But it was just once. I then uploaded new images again and again(doing changes in logging to test), but those tasks were logging only with old logger message(I delivered at first time). Why it was only once that tasks caught the image changes from ECR? What the difference with new images we deliver to ECR? I don't know what causes this behaviour, but it looks strange
2) If it's impossible to tasks re-started with new Docker image, how we can setup task run every 30 seconds or faster? We can change code inside of our image and remove those infinite loop. Then we can setup scheduled tasks in our cluster. That means when we upload new Docker image to ECR, scheduled tasks will run with those image. Looks good, but here is one problem: the minimal time interval there is 1 minute + we will have delay because task needs to prepare resources before start.

**Comment by FernandoMiguel** [comment]
@AndreyAttr you should avoid latest tag, unless you are sure what you do with it.
Ideally you tag them properly, and then update your task definition to use the new label, and then deploy.
If using latest, you can just force new deployment and it will pull latest image.
Or you can kill the task being executed, but that might cause you downtime, not sure that's acceptable.

To change the time it takes to start, you need to look at the registering and deregister time out.
Those are responsible for how quick q new one starts

**Comment by adnxn** [comment]
@AndreyAttr: were you able to find a path forward for this issue? were @FernandoMiguel's suggestions helpful for your use case?

**Comment by adnxn** [comment]
closing this issue since we havent heard back, please feel free to reopen if you need.

**Comment by AndreyAttr** [comment]
@adnxn, 
Yes, we managed to found a solution for our situation.
What @FernandoMiguel has offered is clear, but we really needed our Fargate tasks being able to update with newly uploaded ECR image.
The solution for our problem was:
We marked our image with :latest tag when upload it to ECR. But we caught the event when we upload new image to ECR by writing special rules in CloudTrail. Then we are able to send this even to AWS lambda. And then deliver newly generated STOP MESSAGE to our running Fargate task. Additionally we wrap Fargate task into service instead of running it directly on cluster. And when task get STOP SIGNAL, it finish itself. But because it's running on service, service re-create new task with fresh image from ECR. We solved it this way. So, we still continue to use latest tag, as it allows us to update our newly created task with fresh image all the time. 
But, anyway, big thanks you and @FernandoMiguel for the help.
